impeller/renderer/backend/vulkan/texture_source_vk.h

```diff
@@ -16,53 +16,114 @@
+  ///             mip-counts, types is reliable. So use these descriptors as
+  ///             advisory. Creating copies of texture sources from these
+  ///             descriptors is usually not possible and  depends on the
+  ///             allocator used.
+  ///
+  /// @return     The texture descriptor.
+  ///
   const TextureDescriptor& GetTextureDescriptor() const;
 
+  //----------------------------------------------------------------------------
+  /// @brief      Get the image handle for this texture source.
+  ///
+  /// @return     The image.
+  ///
   virtual vk::Image GetImage() const = 0;
 
-  /// @brief Retrieve the image view used for sampling/blitting/compute with
-  ///        this texture source.
+  //----------------------------------------------------------------------------
+  /// @brief      Retrieve the image view used for sampling/blitting/compute
+  ///             with this texture source.
+  ///
+  /// @return     The image view.
+  ///
   virtual vk::ImageView GetImageView() const = 0;
 
-  /// @brief Retrieve the image view used for render target attachments
-  ///        with this texture source.
+  //----------------------------------------------------------------------------
+  /// @brief      Retrieve the image view used for render target attachments
+  ///             with this texture source.
+  ///
+  ///             ImageViews used as render target attachments cannot have any
+  ///             mip levels. In cases where we want to generate mipmaps with
+  ///             the result of this texture, we need to create multiple image
+  ///             views.
+  ///
+  /// @return     The render target view.
   ///
-  /// ImageViews used as render target attachments cannot have any mip levels.
-  /// In cases where we want to generate mipmaps with the result of this
-  /// texture, we need to create multiple image views.
   virtual vk::ImageView GetRenderTargetView() const = 0;
 
-  /// Encodes the layout transition `barrier` to `barrier.cmd_buffer` for the
-  /// image.
+  //----------------------------------------------------------------------------
+  /// @brief      Encodes the layout transition `barrier` to
+  ///             `barrier.cmd_buffer` for the image.
+  ///
+  ///             The transition is from the layout stored via
+  ///             `SetLayoutWithoutEncoding` to `barrier.new_layout`.
+  ///
+  /// @param[in]  barrier  The barrier.
+  ///
+  /// @return     If the layout transition was successfully made.
   ///
-  /// The transition is from the layout stored via `SetLayoutWithoutEncoding` to
-  /// `barrier.new_layout`.
   fml::Status SetLayout(const BarrierVK& barrier) const;
 
-  /// Store the layout of the image.
+  //----------------------------------------------------------------------------
+  /// @brief      Store the layout of the image.
+  ///
+  ///             This just is bookkeeping on the CPU, to actually set the
+  ///             layout use `SetLayout`.
```

Comment:
technically setLayout doesn't actually set the layout either, it only encodes a command to set the layout. but that won't actually occur until the cmd buffer is submitted. This technically allows layout commands submitted in a different order then they are encoded in to leave the image an an unexpected or invalid state.

The vulkanized presentation on "synchronization made easy" suggests a technique were the state is tracked per queue and updated when the cmd buffer is submited.

Reply:
sgtm. I just moved the previous comments and didn't check for accuracy here.